docs/player.rst

```diff
@@ -0,0 +1,143 @@
+            self._session = session
+
+        @classmethod
+        @contextmanager
+        def create(cls, opponent: "str | None"):
+            """
+            Create a scoped instance of this class and inform it of its ``opponent``
+            """
+            # 'opponent' is the local player, 'session.opponent' the remote player
+            with GameSession.connect(opponent) as session:
+                yield cls(session.opponent, session)
+
+The constructor can then be used in a `with statement`_:
+
+.. code:: python3
+
+    with RemotePlayer.create(...) as player:
+        ...
+
+Note that you can create two players of type ``P1`` and ``P2``
+in one ``with`` statement using
+``with P1.create(...) as player1, P2.create(...) as player2:``.
+
+.. warning::
+
+    The ``opponent`` identifier is sent to whichever player is matched by the server
+    without any authentication.
+    Avoid exposing any personal information such as the account- or hostname
+    (e.g. by creating an ``f"{opponent}@{hostname}"`` identifier).
+    When in doubt, do not provide any identifier - the client API will then
+    create a random one that exposes no information.
+
+Wrapping Methods
+----------------
+
+The methods of :py:class:`~.GameSession` are unlikely to
+directly match ``Player`` methods.
+At the very least, you must *delegate* method calls:
+
+.. code:: python3
+
+    class RemotePlayer:
+        ...
+
+        def get_shot(self):
+            return self._session.expect_shot()
+
+In addition, expect to *adapt* method calls:
+
+.. code:: python3
+
+    class RemotePlayer:
+        ...
+
+        def notify_shot(self, x, y):
+            # adapt the different parameter convention
+            return self._session.announce_shot((y, x))
+
+When expected and provided methods are very different,
+be prepared to implement a *facade* between both conventions:
+
+.. code:: python3
+
+    class RemotePlayer:
+        def __init__(self, ...):
+            ...
+            # translate between individual and all-at-once placements
+            # by storing them internally
+            self._enemy_ship_buffer: "list[SHIP_PLACEMENT] | None" = []
+            self._my_ship_buffer: "list[SHIP_PLACEMENT] | None" = None
+
+        def notify_ship(self, size: int, pos: "tuple[int, int]", vertical: bool):
+            """Inform about enemy placing a ship of specific `size` at `pos`"""
+            # keep collecting all ship placements without sending any
+            self._enemy_ship_cache.append((size, pos, vertical))
+
+        def get_ship(self, size: int) -> "SHIP_PLACEMENT":
+            """Get the next placement for a ship of specific `size`"""
+            # send ship placement only when we need the response
+            if self._my_ship_buffer is None:
+                self._my_ship_buffer = list(
+                    session.place_ships(*self._enemy_ship_cache)
+                )
+                self._enemy_ship_cache = None
```

Comment:
I am confused a bit here. In `__init__` the `_enemy_ship_buffer` is created and in the following the `_enemy_ship_cache` is used. Do I oversee something?

Reply:
Nice catch. I renamed both variables but without IDE support messed that up of course... 😅